src/async_read.rs

```diff
@@ -0,0 +1,80 @@
+use bytes::BufMut;
+use futures01::{Async, Future, IntoFuture, Poll};
+use std::io::Read;
+use tokio01::io::AsyncRead;
+
+/// A AsyncRead combinator which reads from a reader until a future resolves.
+///
+/// This structure is produced by the [`AsyncReadExt::read_until`] method.
+#[derive(Clone, Debug)]
+pub struct ReadUntil<S, F, O> {
+    reader: S,
+    until: F,
+    until_res: Option<O>,
+    free: bool,
+}
+
+/// This `AsyncRead` extension trait provides a `read_until` method that terminates the reader once
+/// the given future resolves.
+pub trait AsyncReadExt: AsyncRead {
+    /// Read data from this reader until the given future resolves.
+    ///
+    /// If the future produces an error, the read will be allowed to continue indefinitely.
+    fn read_until<U, O>(self, until: U) -> ReadUntil<Self, U::Future, O>
```

Comment:
We should probably use another name, this is confusing cause it conflicts with [`futures::io::AsyncBufReadExt::read_until`](https://docs.rs/futures/0.3.5/futures/io/trait.AsyncBufReadExt.html#method.read_until).
The same applies to the `ReadUntil` struct itself.

Reply:
What about `AsyncAllowReadExt` and `allow_read_until`?

Reply:
`allow_read_until` sounds good! Struct name can be simply derived from the method name: `AllowReadUntil` in this case.

I'm assuming `AsyncAllowReadExt` is the proposed new trait name. For the trait, I suggest either use `AllowReadUntilExt` or just `AsyncReadExt` (given it's mod-namespaced it's ok as is). `AsyncReadExt` works better if we plan to all other methods in it, and the more specific `AllowReadUntilExt` works if we use it as an ext for only `allow_read_until` - and we plan on adding other similar traits exts for other things.
If you like `AsyncAllowReadExt` better - I'm fine with that too, the trait name isn't really a concern for me :smile: